### ISSUE TITLE
Respect x as a multiplicator at formulas

### ISSUE THREAD

**Issue by andrcoding** [cgeo/cgeo#13844]
### Tell us your idea!

We have started to discuss it in the past, but I do not remember the final decision.

A lot of formulas in caches use the "x" as a multiplicator. Therefore, it would be helpful if there is an option to parse this formulas correctly. Either by ignoring "x" as a variable or to have an option to replace it while copying/pasting. Or ist this already working with an automatic cache scanning for formulas which I have overlooked the far.

![image](https://user-images.githubusercontent.com/78544594/211195324-f96805a0-57d1-4591-95b6-6d44beb4abe0.png)

### What solution would you suggest?

see above

**Comment by eddiemuc** [comment]
This is a reoccurring topic. I guess we could find a special handling here, but probably then the users can't use x as variable name any more

**Comment by ztNFny** [comment]
This (recognize x as *) could be implemented in the scanner, then calculation can stay as it is.
Handling x as *and therefore removing any other use of x send like a very bad idea

**Comment by eddiemuc** [comment]
> This (recognize x as *) could be implemented in the scanner, then calculation can stay as it is.
> Handling x as *and therefore removing any other use of x send like a very bad idea 

Iirc replacement of x with * **is** already implemented in scanner. But if users enter formulas directly or using "normal" copy/paste then this does not help

**Comment by ztNFny** [comment]
> > This (recognize x as *) could be implemented in the scanner, then calculation can stay as it is.
> > Handling x as *and therefore removing any other use of x send like a very bad idea 
> 
> Iirc replacement of x with * **is** already implemented in scanner. But if users enter formulas directly or using "normal" copy/paste then this does not help

Then it's suggest to close here. Input must always be interpreted as entered. Shit in, shit out

**Comment by andrcoding** [comment]
You're right. The scanner is working correctly. So far I know, the scanner is limited to the variable view, isn't it? In this case it would be helpful to have the scanner also available at the waypoint/coordinate calculation or can I use just one variable for one of the coordinates?

**Comment by eddiemuc** [comment]
I think I found a solution for this: we can interpret lower-case x as multiply symbol (not as variable) if is is surrounded by whitespaces. E.g. in case of a variable x defined with value 2, then following would be interpreted:
* `1x3` would be `123`
* `1 x 3` would be `1 * 3` = `3`

There are more corner cases, but this is doable. I will push a PR soon

**Comment by bekuno** [comment]
> I think I found a solution for this: we can interpret lower-case x as multiply symbol (not as variable) if is is surrounded by whitespaces.

This is a good idea.
But often there is instead a brace.
So instead checking for a white space lets check for a non alphabetic symbol.

**Comment by bekuno** [comment]
Or as in your sample for a non alphanumeric symbol.

**Comment by ztNFny** [comment]
> I think I found a solution for this: we can interpret lower-case x as multiply symbol (not as variable) if is is surrounded by whitespaces. E.g. in case of a variable x defined with value 2, then following would be interpreted:
> * `1x3` would be `123`
> * `1 x 3` would be `1 * 3` = `3`
> 
> There are more corner cases, but this is doable. I will push a PR soon

Not really convinced of that. I'd leave it as is and only do the magic in the recognition part.

**Comment by eddiemuc** [comment]
> > I think I found a solution for this: we can interpret lower-case x as multiply symbol (not as variable) if is is surrounded by whitespaces.
> 
> This is a good idea. But often there is instead a brace. So instead checking for a white space lets check for a non alphabetic symbol.

No, I feel that concatenation should still work in this case. E.g. (for `x=2`):
* `3x(1+2)` should still be `323`
* `3 x (1+2)` should be `9`

**Comment by eddiemuc** [comment]
> Not really convinced of that. I'd leave it as is and only do the magic in the recognition part.

Why are you not convinced?

**Comment by ztNFny** [comment]
> > Not really convinced of that. I'd leave it as is and only do the magic in the recognition part.
> 
> Why are you not convinced? 

Interpreting a symbol differently depending on context will lead to confusion. It'll be near impossible for someone who doesn't know why axb and a x b are behaving differently, especial as according to all mathematical rules whitespace doesn't matter.
1 char should do 1 thing

**Comment by eddiemuc** [comment]
> Interpreting a symbol differently depending on context will lead to confusion. It'll be near impossible for someone who doesn't know why axb and a x b are behaving differently

Correct. The argument would be that it works almost ever time as expected and otherwise the manual has to be consulted.

> especial as according to all mathematical rules whitespace doesn't matter. 

Also correct. But if we take mathematical rules seriously then we can't allow concatenation at all (something like `2AB` must be interpreted as `2*A*B`). We don't want that and apply "geocaching-mathematics" instead. Whether the `x`-case is also "geocaching-mathematics" or not is discussable of course.

**Comment by ztNFny** [comment]
As there's no correct answer on this but it depends on whatever the cache author felt like doing - he could even use the • sign as multiplication char - my vote is for a clearly rule-following and thus to the user understandable solution - which is x is a letter, just like y and z.

**Comment by ztNFny** [comment]
In terms of user expectation: when i get a formula from a listing i know that there's plenty of ways to write things so whatever cgeo auto-recognizes i will make sure it's what i expect.
But when i enter a formula i want it to be parsed exactly as entered, no "magic" should happen there. I'll never expect x to be something other than x - and even more so if it works sometimes and only fails in a few cases (depending on whitespace), this will 100% make me walk to wrong coords as i'll trust the output as it's resulting from a formula i personally checked and have no way of veryfing again without doing the calculation manually - in which case cgeo becomes completely useless if i can't trust it and have to double-check everything

**Comment by ztNFny** [comment]
In terms of user expectation: when i get a formula from a listing i know that there's plenty of ways to write things so whatever cgeo auto-recognizes i will make sure it's what i expect.
But when i enter a formula i want it to be parsed exactly as entered, no "magic" should happen there. I'll never expect x to be something other than x - and even more so if it works sometimes and only fails in a few cases (depending on whitespace), this will 100% make me walk to wrong coords as i'll trust the output as it's resulting from a formula i personally checked and have no way of veryfing again without doing the calculation manually - in which case cgeo becomes completely useless if i can't trust it and have to double-check everything

**Comment by eddiemuc** [comment]
Many Geocachers are not programmers and thus not used to * and / as mathematical symbols. I assume this group of Geocache owners is also the one using x and : in their coordinate formulas.

Anyway, there are always other options:
* Leave everything as is
* Forbid using x as a variable
* Introduce a setting where the user can turn off the behavior implemented in the PR (so c:geo becomes usable and trustworthy again)

**Comment by andrcoding** [comment]
I think there is no solution for everybody. After this discussion I have a better understanding of your thoughts and can live with both concepts. 
So far, you have fortunately tried to keep the options menu relatively slim. If necessary, you could also add an option for interpretation, but this would blow up the code. As you mentioned.

PS: Actually the only correct mathematical symbol for multiplication is "·" and not "*" ;)

**Comment by ztNFny** [comment]
> PS: Actually the only correct mathematical symbol for multiplication is "·" and not "*" ;)

Nope :-P
× is multiplication, ⋅ is dot-product (which for non-vectors results in the same as ×)

**Comment by ztNFny** [comment]
> Anyway, there are always other options:

I'd like that in the end we always have a field with an unambiguous formula which gives THE expected result.
To get to that formula requires some "magic" to recognize what the cache owner (or even the user) might have meant.

To do that how about a button next to every field that's capable of doing formulas to run formula recognition on it? In addition we could indicate to the user whether the field currently contains a valid formula and point him to the auto-recognition:
![image](https://user-images.githubusercontent.com/1258173/212553199-d61e3367-cdb8-46e4-a04a-e6c8883a44ec.png)

When clicking the button the user is presented with a parsed formula which can be either be applied or canceled. And if we feel fancy we can even present multiple options for the user to select there.
E.g. for input "1 x 2" the user would get a dialog:

[ ] 1*2
[ ] 1x2
[Apply] [Cancel]

**Comment by eddiemuc** [comment]
> > Anyway, there are always other options:
> 
> I'd like that in the end we always have a field with an unambiguous formula which gives THE expected result. To get to that formula requires some "magic" to recognize what the cache owner (or even the user) might have meant.
> 
> To do that how about a button next to every field that's capable of doing formulas to run formula recognition on it? In addition we could indicate to the user whether the field currently contains a valid formula and point him to the auto-recognition: 

I might not understand the idea correct. 
* The current formula parsing algorithm is and must remain unambigious - with and without the x-multiplier-PR applied.
* If it does not recognize a formula then the var tab already gives a message indicating what it thinks is wrong. 

Here is an example screenshot of 4 formulas (one formula is correct, the others are not. Screenshot was done against a build with x-multiplier-PR applied, variable x is not defined):

![image](https://user-images.githubusercontent.com/6909759/212556695-110fdf86-e473-422a-9138-642eb31b5ea9.png)

Granted, we can talk in lengths about how this validation message can be enhanced (there was no discussion about this before) but it is there in principle.

**Comment by ztNFny** [comment]
What's not there is the "apply magic to input" button and the message for the "maybe-wrong" case

**Comment by eddiemuc** [comment]
> What's not there is the "apply magic to input" button (which is where i'd suggest to put your "x as *"-recognition) and the message for the "maybe-wrong" case

I wouldn't know how to implement those. Since parsing is unambigious, by which rules should "maybe-wrong" cases be identified? To the parser everything applying to its rules is right, everything it can't parse is wrong. Something like `1x5` or `1 x 5` can both be parsed unambigious, be it with or without the PR applied.

And which "magic" should be applied to which inputs under which conditions?

**Comment by ztNFny** [comment]
> And which "magic" should be applied to which inputs under which conditions?

E.g. the "magic" your PR wants to apply: treat "x" differently depending on input.
And the already implemented logic for listing scanning, e.g. recognizing "·" as "*".

> Something like `1x5` or `1 x 5` can both be parsed unambigious, be it with or without the PR applied.

That's why I suggested to present the users with options he might or might not accept:
> When clicking the button the user is presented with a parsed formula which can be either be applied or canceled. And if we feel fancy we can even present multiple options for the user to select there:
> ![image](https://user-images.githubusercontent.com/1258173/212553404-e482c571-431a-402b-9a8c-0195aebfbcb8.png)

**Comment by ztNFny** [comment]
It seems you're mainly focusing on the "maybe-wrong" note. My main point is: All "intelligent" parsing MUST NOT happen behind the scenes but with an explicit "apply magic" button next to the text field.
The text field must be parsed exactly as is - meaning only "+\-*/%^*:!" (VALID_OPERATOR_PATTERN) and the defined functions (like letter-value and so on) are valid operators, everything else - e.g. the "x" - is a variable.

![image](https://user-images.githubusercontent.com/1258173/212558376-ac537804-375d-4a1c-bff0-62d5bbaf687e.png)

**Comment by eddiemuc** [comment]
@ztNFny I think we agree about the fact that parsing must remain (and is) unambigious at all times. That is: there must be fixed rules applied to a given string on how the chars in it have to be interpreted. Those rules must be documented. Also, those rules must be intuitive so the user does not interpret a written formula different on what the parser does. And they must fit to the geocaching world and to the typical geocaching user.

I think where we differ is what rules should be considered "best" for typical geocaching users of c:geo.

I believe that we should NOT accept just one dedicated sign for each mathematical operation. I also don't believe that the signs taken in programming languages are the most intuitive ones for geocachers. That's why I argue to accept a broad range of symbols as long as everything is unambigious. For example, I see not reason not to accept `:` or `÷` in addition to `/` as divison operator (I believe you agree with `:` but not with `÷`). By the same argumentation I would accept `—` for minus and `•` for multiplication. 

Problems arise if symbols are used ambigious in other contexts. Examples in c:geo:
* `x` can be seen as variable name but also as multiplication symbol
* `%` can be seen as modulo operator but also as percentage operator

For those cases we have to find individual rules which best fit to the intuitive usage scnearios for most geocachers. I believe that many geocachers use `x` as multiplicator, and I want to find rules which accomodate to this. I believe the "whitespace before/after" rule is a good compromise. But that's just my opinion.

All in all: I believe there is not one "universal intelligent" rule set which symbols to apply for which operation. I also believe that usage of programming symbols or strict mathematical rules is not what most geocachers would find intuitive. So I struggle to find the best compromise for the majority of c:geo users out there. 

Maybe this is something for the next team meeting?

**Comment by ztNFny** [comment]
> would find intuitive

that's the main point here - and one that IMHO fails with your implementation.

If you show "xx x x; x=2" to anyone I'm pretty sure there's no expectation that this will be parsed to 44 and "x x x x" as ... actually I have no idea what that will be parsed as!

Also I don't think anyone would expect "1x3" and "1 x 3" to yield different results.

That's why - again IMHO - your PR is bad in terms of UX.


My counter-proposal is to offer the user a mechanism that enables easy application of formulas by suggesting possible interpretations of that formula to the user, e.g. for "xx x x" to offer "xx*x". When the user sees that formula it'll will be very clear to him what the result is, no ambiguity at all there.

**Comment by bekuno** [comment]
> If you show "xx x x; x=2" to anyone I'm pretty sure there's no expectation that this will be parsed to 44 and "x x x x" as 

Sorry, that sample is crap.
I believe that no owner will write this.
Maybe he would write "XX x X; X=2". 
But I hope that no owner will use X and x in a formula.

This brings me to a new proposal:
For recognizing `x` (lower case x) as `*`, all other variables have to be in upper case.

**Comment by eddiemuc** [comment]
Let's end this. I believe that in practice my proposal would solve the vast majority of real-life use cases while at the same time being a very small and structured change. Sure one could further philosophe about the case of `1x3` vs `1 x 3`, I believe this has no relevance in practice.

Examples like `xx x x` are for nerds like me, they have no practical relevance imho.

If a more complex solution is wanted which involves guessing what the user wants and asking him to check and correct what he/she wants - anyone may feel free to propose a fine-tuned UX and implement it. Approaching this path feels too complex to me for such a simple problem.

Interpreting x as multiplier if there are only uppercase vars is not easy to implement. It would at least involve parsing every formula twice (once to find out which vars are in, a second time to re-interpret the x). I don't stand for it.

I will leave my PR open for the time being and rest my case.

**Comment by Lineflyer** [comment]
My few cents after reading through all of it:

The x versus * is a "problem" in used formulas of geocaching. However I do only see the x being used rarely as IMHO most cache owners refrain from using the letter x even due to its multiple interpretations.

That being said, I also think that differentiating `1x3` from `1 x 3` is really not to be expected by any user without further explanation.
Of course there could be some information for or decision by the user....but to be honest. I would vote for leaving it as it is right now by not treating x as a mathematical operator after all.

You could also phrase it in another way:
We will fail to interpret all kind of formulas in a correct way no matter how hard we try as long as there is no common coding. If we losen our syntax it might lead to false results or even unexpected chain-failures.

In such cases, where the owner uses `x` for multiplication I might instead friendly ask him (if its of common use) to use `*` instead and - with doing this - make his fomulas compatible to most tools out there.

**Comment by ztNFny** [comment]
Some inputs from the duplicate:

              Second thought:
We could think about having a new function in the "magic wand menu" like "Interpret x as multiplication" to on demand auto search and replace all occurences of `x` bx `*` in a formula, which has been entered. Would that be a feasible compromise for such scenarios?

_Originally posted by @Lineflyer in https://github.com/cgeo/cgeo/issues/14461#issuecomment-1657230858_


              > I agree to the "shit in shit out" Mantra in general. However, I feel this issue is more about what to consider "shitty" and what not. E.g. is the interpretation of a formula `3 x 4` as `3 times 4` shitty or not.

Completely agreed.

> (or other formal ruleset,)

As I already said, my rather simple ruleset would be: 

```
Ignoring spaces, if x is enclosed by capital letters and/or parantheses, replace it by `*`.
(Here the N/S and E/W in the relevant places of coordinates are not considered capital letters.)

Otherwise, treat it as a variable.
```

Not very advanced, but already this simple rule handles a lot more caches correctly than the current one.

_Originally posted by @dcjkfgdjhd in https://github.com/cgeo/cgeo/issues/14461#issuecomment-1660069667_

**Comment by ztNFny** [comment]
My personal preference at this point would be to not implement any "magic" but go with @Lineflyer suggestion and add a button/menu item that does a simple "replace("x", "*")" when the user actively decides to do so.
